src/main/java/org/mule/extension/api/impl/introspection/ImmutableExtensionModel.java

```diff
@@ -0,0 +1,226 @@
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<OperationModel> getOperationModels()
+    {
+        return toList(operations.values());
+    }
+
+    @Override
+    public List<SourceModel> getSourceModels()
+    {
+        return toList(messageSources.values());
+    }
+
+    @Override
+    public SourceModel getSourceModel(String name) throws NoSuchMessageSourceException
+    {
+        return findModel(messageSources, name, () -> new NoSuchMessageSourceException(this, name));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getVersion()
+    {
+        return version;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OperationModel getOperationModel(String name) throws NoSuchOperationException
+    {
+        return findModel(operations, name, () -> new NoSuchOperationException(this, name));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<ConnectionProviderModel> getConnectionProviders()
+    {
+        return connectionProviders;
+    }
+
+    private <T extends EnrichableModel> T findModel(Map<String, T> map, String name, Supplier<RuntimeException> exception)
+    {
+        T model = map.get(name);
+        if (model == null)
+        {
+            throw exception.get();
+        }
+
+        return model;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getVendor()
+    {
+        return vendor;
+    }
+
+    @Override
+    public Optional<ExceptionEnricherFactory> getExceptionEnricherFactory()
+    {
+        return exceptionEnricherFactory;
+    }
+
+    @Override
+    public String toString()
+    {
+        return "ImmutableExtensionModel{" +
+               super.toString() +
+               ", version='" + version + '\'' +
+               ", configurations=" + configurations +
+               ", operations=" + operations +
+               ", connectionProviders=" + connectionProviders +
+               ", messageSources=" + messageSources +
+               ", vendor='" + vendor + '\'' +
+               ", exceptionEnricherFactory=" + exceptionEnricherFactory +
+               '}';
+    }
+
+    private <T extends Described> Map<String, T> toMap(List<T> objects)
+    {
+        Map<String, T> map = new LinkedHashMap<String, T>();
+        for (T object : objects)
+        {
+            if (map.containsKey(object.getName()))
+            {
+                throw new IllegalArgumentException(String.format("Multiple entries with the same key[%s]", object.getName()));
+            }
+            map.put(object.getName(), object);
+        }
+        return Collections.unmodifiableMap(map);
+    }
+
+    private <T extends Described> List<T> toList(Collection<T> collection)
+    {
+        if (collection == null || collection.isEmpty())
+        {
+            return Collections.emptyList();
+        }
+
+        List<T> list = new ArrayList<T>();
+        list.addAll(collection);
```

Comment:
don't do addAll(). Directly pass the collection into the new list's constructor and inline it with the return statement below